lighthouse-core/audits/seo/font-size.js

```diff
@@ -121,80 +120,84 @@ function nodeToTableNode(node) {
-          // is inline.
-          // Always use the rule's location if a sourceURL magic comment is
-          // present (`hasSourceURL` is true) - this makes the line/col relative to the start
-          // of the style tag, which makes them relevant when the "file" is open in DevTool's
-          // Sources panel.
-          const addHtmlLocationOffset = stylesheet.isInline && !stylesheet.hasSourceURL;
-          if (addHtmlLocationOffset) {
-            line += stylesheet.startLine;
-            // The column the stylesheet begins on is only relevant if the rule is declared on the same line.
-            if (range.startLine === 0) {
-              column += stylesheet.startColumn;
-            }
-          }
-
-          source += `:${line}:${column}`;
-        }
-      } else {
-        // dynamically injected to page
-        source = 'dynamic';
-      }
+    selector = rule.selectors.map(item => item.text).join(', ');
+  }
 
-      return {
-        selector,
-        source,
-      };
+  if (styleDeclaration.stylesheet && !styleDeclaration.stylesheet.sourceURL) {
+    // Dynamically injected into page.
+    return {
+      source: {type: 'code', value: 'dynamic'},
+      selector,
+    };
+  }
+
+  // !!range == has defined location in a source file (.css or .html)
+  // sourceURL == stylesheet URL || raw value of magic `sourceURL` comment
+  // hasSourceURL == flag that signals sourceURL is the raw value of a magic `sourceURL` comment, *not* a real resource
+  if (styleDeclaration.stylesheet && styleDeclaration.range) {
+    const {range, stylesheet} = styleDeclaration;
+
+    // DevTools protocol does not provide the resource URL if there is a magic `sourceURL` comment.
+    // `sourceURL` will be the raw value of the magic `sourceURL` comment, which likely refers to
+    // a file at build time, not one that is served over the network that we could link to.
+    const urlIsNetworkResource = !stylesheet.hasSourceURL;
+
+    let line = range.startLine;
+    let column = range.startColumn;
+
+    // Add the startLine/startColumn of the <style> element to the range, if stylesheet
+    // is inline.
+    // Always use the rule's location if a sourceURL magic comment is
+    // present (`hasSourceURL` is true) - this makes the line/col relative to the start
+    // of the style tag, which makes them relevant when the "file" is open in DevTool's
+    // Sources panel.
+    const addHtmlLocationOffset = stylesheet.isInline && !stylesheet.hasSourceURL;
+    if (addHtmlLocationOffset) {
+      line += stylesheet.startLine;
+      // The column the stylesheet begins on is only relevant if the rule is declared on the same line.
+      if (range.startLine === 0) {
+        column += stylesheet.startColumn;
+      }
     }
+
+    const url = stylesheet.sourceURL;
+    return {
+      source: {type: 'source-location', urlIsNetworkResource, url, line, column},
```

Comment:
i think we should migrate a second audit too.. like `deprecations` or `errors-in-console`.  i'm curious if any of this line adjustments would move to a common file.
(fwiw devtools has some utility methods for this kinda stuff... `toUILocation()` iirc)

Reply:
Started with deprecations. every one i looked at seemed to not have any line number or url. got pretty confused. this [comment](https://github.com/GoogleChrome/lighthouse/commit/fadaef95137181aa41ed44fd78f17d3a8b85a2c3#diff-47bc5ba5cda5a06117ea0ff99fefe0c1L42) (lost to a refactor) clears it up: only css deprecations have url/lineNumber.

Unknown to me if there are JS deprecations too.

Reply:
Started with deprecations. each one i looked at seemed to not have any line number or url. got pretty confused.

this [comment](https://github.com/GoogleChrome/lighthouse/commit/fadaef95137181aa41ed44fd78f17d3a8b85a2c3#diff-47bc5ba5cda5a06117ea0ff99fefe0c1L42) (lost to a refactor) was useful, links me to a bug where it's mentioned that this deprecation system was refactored.

> // CSS deprecations can have missing URLs and lineNumbers. See https://crbug.com/680832.

I'm guessing the HTML deprecations never have a url/line number, the CSS ones do, and the JS ones (are there any?) should too.

Reply:
nvm figured it out.